add docker image integration test

xvello opened the pull request:
Add a new docker:integration_test rake command to run an integration check on the dogstatsd alpine docker image.
Builds and runs the image both in UDP and socket mode. With lsof, we test that dogstatsd is running and listening on the right interface.

Written in shell for convenience, can be ported to Go once we add the docker go library to the build deps.

Didn't add a docker:build command as we'd need to define if/how to remove the built image after the run. The test builds its own image and deletes it on exit.

For now, the test downloads the docker client if not found in the path, will modify the debian runner to install docker in it, then remove that branch in the test

Modified the pipeline to:
source tests -> binary build -> integration tests -> package builds -> deploy

We could add an end_to_end_tests stage between package_build and deploy once we have one

xvello committed:
add docker image system test

xvello committed:
exit docker:system_test if script fails

xvello committed:
fix dogstatsd:system_test exit code

xvello committed:
tentative fix to run on gitlab

xvello committed:
s/system/integration/

xvello committed:
update comments

xvello committed:
rework pipeline to delay package builds after int. tests

masci reviewed (approved):
LGTM!

dbenamydd reviewed (commented):
This is a great first pass!!

We might want to run this using docker-in-docker if that's not too hard. Right now we have shell runners that work the way you've used them, but we're trying to move away from them for security isolation. There are some cases where this is hard, but this might be an easy one!

> Didn't add a docker:build command as we'd need to define if/how to remove the built image after the run. The test builds its own image and deletes it on exit.

Is this referring to the `dsd-alpine` image?

You probably don't need to worry about removing it. The runners have a cron job to clean up old docker images.

xvello committed:
docker integration test simplification

- move the image creation to the docker:build_dogstatsd task
- remove docker installation as the runners will have it
- don't bind the socket out of the container (not needed for the test)
- update .gitignore

gmmeyer reviewed (approved):
I have two small comments. One is a nit, one is a question. However, lgtm as is.
